// @acoss/prisme-angular-intranet/esm2020/lib/component/callback/callback.component.mjs
import { Component } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../prisme-angular.config";
import * as i2 from "../../prisme-angular-init.config";
import * as i3 from "@angular/router";
export class CallbackComponent {
    constructor(environment, prismeAngularInitConfig, router) {
        this.environment = environment;
        this.prismeAngularInitConfig = prismeAngularInitConfig;
        this.router = router;
    }
    ngOnInit() {
        // Récupération du paramètre pageFrom dans le sessionStorage (le pageFrom n'est pas présent lorsqu'on provient d'un portail)
        const pageFrom = sessionStorage.getItem(this.prismeAngularInitConfig.cleStockagePageFrom);
        // Si pageFrom est null alors on provient d'un portail --> pas besoin de redirection
        if (pageFrom) {
            // Redirection vers la page pageFrom
            this.router.navigate([pageFrom]);
        }
    }
}
CallbackComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: CallbackComponent, deps: [{ token: i1.PrismeAngularConfiguration }, { token: i2.PrismeAngularInitConfig }, { token: i3.Router }], target: i0.ɵɵFactoryTarget.Component });
CallbackComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.1.0", type: CallbackComponent, selector: "lib-callback", ngImport: i0, template: "<p>\n   R&eacute;cup&eacute;ration des infos de connexion en cours...\n</p>\n" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: CallbackComponent, decorators: [{
            type: Component,
            args: [{ selector: 'lib-callback', template: "<p>\n   R&eacute;cup&eacute;ration des infos de connexion en cours...\n</p>\n" }]
        }], ctorParameters: function () { return [{ type: i1.PrismeAngularConfiguration }, { type: i2.PrismeAngularInitConfig }, { type: i3.Router }]; } });
