// @acoss/prisme-angular-intranet/esm2020/lib/service/access-token.service.mjs
import { Injectable } from '@angular/core';
import * as pako from 'pako';
import { JwtHelperService } from '@auth0/angular-jwt';
import * as i0 from "@angular/core";
/**
 * Service permettant de décoder un access token et d'en extraire les informations
 */
export class AccessTokenService {
    constructor() {
        this.jwtHelper = new JwtHelperService();
    }
    getIdentifiantUtilisateur(accessToken) {
        return this.decodeAccessToken(accessToken).sub.split('@')[0];
    }
    getDroitsUtilisateur(accessToken) {
        // La liste des droits est située dans le "aud" du jeton d'accès au format base64
        const decode_aud = atob(this.decodeAccessToken(accessToken).aud);
        // Après décodage, la chaine est gzippée : utilisation de la bibliothèque pako pour décompresser
        const droits_string = pako.inflate(decode_aud, { to: 'string' });
        // Les droits FRONT sont séparés par un espace
        return droits_string.split(' ');
    }
    decodeAccessToken(accessToken) {
        return this.jwtHelper.decodeToken(accessToken);
    }
    getTokenExpirationDate(accessToken) {
        return this.jwtHelper.getTokenExpirationDate(accessToken);
    }
    isTokenExpired(accessToken) {
        return this.jwtHelper.isTokenExpired(accessToken);
    }
}
AccessTokenService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: AccessTokenService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
AccessTokenService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: AccessTokenService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: AccessTokenService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return []; } });
