// @acoss/prisme-angular-intranet/esm2020/lib/service/login.service.mjs
import { Inject, Injectable } from '@angular/core';
import { PrismeAngularInitConfig } from '../prisme-angular-init.config';
import { InfosUtilisateur } from '../model/infos-utilisateur';
import * as i0 from "@angular/core";
import * as i1 from "./access-token.service";
import * as i2 from "../prisme-angular-init.config";
export class LoginService {
    constructor(config, accessTokenService) {
        this.config = config;
        this.accessTokenService = accessTokenService;
    }
    /**
     * Indique si l'utilisateur est connecté.
     */
    isConnecte() {
        const logInfo = this.getInfos();
        return logInfo !== null && logInfo.accessTokenBack !== null;
    }
    /**
     * Renvoie toutes les informations
     */
    getInfos() {
        const infos = this.getStorage().getItem(this.config.cleStockageInfosUtilisateur);
        if (infos === null) {
            return null;
        }
        return JSON.parse(infos);
    }
    /**
     * Renvoie les infos de l'utilisateur
     */
    getInfosUtilisateurBack() {
        const infos = this.getInfos();
        if (infos != null && infos.infosUtilisateurBack != null) {
            return new InfosUtilisateur(infos.infosUtilisateurBack);
        }
        return null;
    }
    getInfosUtilisateurFront() {
        const infos = this.getInfos();
        if (infos != null && infos.infosUtilisateurFront != null) {
            return new InfosUtilisateur(infos.infosUtilisateurFront);
        }
        return null;
    }
    /**
     * Renvoie le payload
     */
    getPayload() {
        const infos = this.getInfos();
        if (infos === null) {
            return null;
        }
        return infos.payload;
    }
    /**
     * Renvoie les droits de l'utilisateur pour le frontend
     */
    getDroitsUtilisateur() {
        const sessionUtilisateur = this.getInfos();
        if (sessionUtilisateur === null) {
            return null;
        }
        return this.accessTokenService.getDroitsUtilisateur(sessionUtilisateur.accessTokenFront);
    }
    /**
     * Renvoie les droits de l'utilisateur pour le backend
     */
    getDroitsUtilisateurBack() {
        const sessionUtilisateur = this.getInfos();
        if (sessionUtilisateur === null) {
            return null;
        }
        return this.accessTokenService.getDroitsUtilisateur(sessionUtilisateur.accessTokenBack);
    }
    getAccessTokenBack() {
        const infos = this.getInfos();
        if (infos === null) {
            return null;
        }
        return infos.accessTokenBack;
    }
    getAccessTokenFront() {
        const infos = this.getInfos();
        if (infos === null) {
            return null;
        }
        return infos.accessTokenFront;
    }
    getOceanToken() {
        const infos = this.getInfos();
        if (infos === null) {
            return null;
        }
        return infos.oceanToken;
    }
    getMode() {
        const infos = this.getInfos();
        if (infos === null) {
            return null;
        }
        return infos.mode;
    }
    /**
     * Enregistre les infos utilisateur
     * @param infos Jeton
     */
    storeInfos(infos) {
        // On récupère le payload existant
        let oldinfos;
        oldinfos = JSON.parse(this.getStorage().getItem(this.config.cleStockageInfosUtilisateur));
        if (oldinfos && infos) {
            infos.payload = infos.payload || oldinfos.payload;
        }
        this.getStorage().setItem(this.config.cleStockageInfosUtilisateur, JSON.stringify(infos));
    }
    /**
     * Pour le mode portail, permet de mettre à jour l'access token dans les infos utilisateur.
     * @param accessToken Jeton
     * @param expiration Date
     */
    updateAccessToken(accessToken, expiration) {
        const infos = this.getInfos();
        if (infos !== null) {
            infos.accessTokenBack = accessToken;
            infos.expiration = expiration;
        }
        this.storeInfos(infos);
    }
    /**
     * Permet de mettre à jour les infos utilisateur.
     * @param infosUtilisateur les infos utilisateur à mettre à jour
     */
    updateInfosUtilisateurBack(infosUtilisateur) {
        const infos = this.getInfos();
        if (infos !== null) {
            infos.infosUtilisateurBack = infosUtilisateur;
            this.storeInfos(infos);
        }
    }
    /**
     * Permet de mettre à jour les infos utilisateur.
     * @param infosUtilisateur les infos utilisateur à mettre à jour
     */
    updateInfosUtilisateurFront(infosUtilisateur) {
        const infos = this.getInfos();
        if (infos !== null) {
            infos.infosUtilisateurFront = infosUtilisateur;
            this.storeInfos(infos);
        }
    }
    getIdentifiantUtilisateur() {
        const sessionUtilisateur = this.getInfos();
        return this.accessTokenService.getIdentifiantUtilisateur(sessionUtilisateur.accessTokenFront);
    }
    /**
   * Permet de mettre à jour le payload
   * @param payload a mettre à jour
   */
    updatePayload(payload) {
        const infos = this.getInfos();
        if (infos !== null) {
            infos.payload = payload;
            this.storeInfos(infos);
        }
    }
    /**
     * Controle si les données utilisateur contiennent bien les droits demandés
     * @param droits Liste des droits
     */
    estAutorise(droits) {
        return this.checkAutorisation(droits, this.getDroitsUtilisateur());
    }
    /**
     * Controle si les données utilisateur contiennent bien les droits demandés pour le backend
     * @param droits Liste des droits
     */
    estAutoriseBack(droits) {
        return this.checkAutorisation(droits, this.getDroitsUtilisateurBack());
    }
    checkAutorisation(droits, droitsJeton) {
        if (!this.isConnecte()) {
            return false;
        }
        for (const droit of droits) {
            if (this.controleAutorisation(droit, droitsJeton)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Controle si les données utilisateur contiennent bien le droit spécifique demandé
     * On considère que les droits sont de la forme <perimetre>:<application>:<profil>
     * @param droit Droit
     * @param droitsJeton la liste des droits contenu dans le jeton
     */
    controleAutorisation(droit, droitsJeton) {
        if (!droitsJeton) {
            return false;
        }
        // Si la chaine <perimetre>:<application>:<profil> contient le droit, alors l'utilisateur est autorisé
        for (const droitJeton of droitsJeton) {
            const values = droitJeton.split(':');
            if (values && -1 !== values.indexOf(droit)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Déconnecter
     * @returns Promise<any> renvoi resolve(true) lorsque la déconnexion est terminée
     */
    deconnecter() {
        return new Promise((resolve, reject) => {
            this.getStorage().clear();
            localStorage.clear();
            resolve(true);
        });
    }
    getStorage() {
        if (this.config.utiliserLocalStorage) {
            return localStorage;
        }
        return sessionStorage;
    }
}
LoginService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: LoginService, deps: [{ token: PrismeAngularInitConfig }, { token: i1.AccessTokenService }], target: i0.ɵɵFactoryTarget.Injectable });
LoginService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: LoginService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: LoginService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i2.PrismeAngularInitConfig, decorators: [{
                    type: Inject,
                    args: [PrismeAngularInitConfig]
                }] }, { type: i1.AccessTokenService }]; } });
