// @acoss/prisme-angular-intranet/esm2020/lib/service/oauth-validation.service.mjs
import { Injectable, Inject } from '@angular/core';
import * as jsrsasign from 'jsrsasign';
import { PrismeAngularConfiguration } from '../prisme-angular.config';
import * as i0 from "@angular/core";
import * as i1 from "../prisme-angular.config";
export class OauthValidationService {
    constructor(environment) {
        this.environment = environment;
    }
    /**
     * Valider que le champ "iss" du token correspond bien au provider
     * "Malicious compliance" vis à vis du guide développeur
     * @param token Jeton
     */
    validerIss(token) {
        return token.iss === this.environment.prismeIss;
    }
    /**
     * Valider que le champ "exp" du token ne correspond pas à une date passée
     * @param token Jeton
     */
    validerExp(token) {
        // Construire un objet Date qui contient la date d'expiration du jeton
        const dateExp = new Date(0);
        dateExp.setUTCSeconds(token.exp);
        if (!dateExp) {
            return false;
        }
        // Comparer la date d'expiration à la date courante
        return dateExp.valueOf() > new Date().valueOf();
    }
    /**
     * Valider que le token est correctement signé
     * @param token Jeton
     * @param keys Cles
     * @param alg Algorithme
     */
    validerSignature(token, keys, alg) {
        // Récupérer le header du jeton
        const header_jeton_encode = token.split('.')[0];
        const header_decode = JSON.parse(atob(header_jeton_encode));
        // L'algorithme doit etre R256
        if (header_decode.alg !== alg) {
            return false;
        }
        let resultat = false;
        // Si une seule clé présente, pas de controle du kid
        if (keys.length === 1) {
            const key = keys[0];
            const clePublique = jsrsasign.KEYUTIL.getKey(key);
            resultat = jsrsasign.KJUR.jws.JWS.verify(token, clePublique, [alg]);
            return resultat;
        }
        // Comparaison de chaque clé à la clé fournie dans le jeton
        for (const key of keys) {
            if (key.kid === header_decode.kid) {
                const clePublique = jsrsasign.KEYUTIL.getKey(key);
                resultat = jsrsasign.KJUR.jws.JWS.verify(token, clePublique, [alg]);
                return resultat;
            }
        }
        return resultat;
    }
    /**
     * Valider que le champ "aud" du token n'est pas vide
     * @param token Jeton
     */
    validerAudNonVide(token) {
        return token.aud && token.aud.length > 0;
    }
    /**
   * Valider que le champ "sub" du token est égal au sujet + '@' + le client ID requis
   * @param token Jeton
   * @param sujet Sujet
   * @param client_id Identifiant
   */
    validerSubOcean(token, client_id) {
        const sub = token.sub.split('@');
        return sub.length === 2 && sub[1] === client_id;
    }
}
OauthValidationService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: OauthValidationService, deps: [{ token: PrismeAngularConfiguration }], target: i0.ɵɵFactoryTarget.Injectable });
OauthValidationService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: OauthValidationService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: OauthValidationService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.PrismeAngularConfiguration, decorators: [{
                    type: Inject,
                    args: [PrismeAngularConfiguration]
                }] }]; } });
