// @acoss/prisme-angular-intranet/esm2020/lib/handler/default-success-login.handler.mjs
import { Inject, Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { PrismeAngularConfiguration } from '../prisme-angular.config';
import { AbstractSuccessLoginHandler } from './abstract-success-login.handler';
import * as i0 from "@angular/core";
import * as i1 from "../prisme-angular-init.config";
import * as i2 from "../prisme-angular.config";
export class DefaultSuccessLoginHandler extends AbstractSuccessLoginHandler {
    constructor(environment, initConfig) {
        // Appel du constrcuteur du parent
        super();
        this.environment = environment;
        this.initConfig = initConfig;
        // Test des paramètres obligatoires
        if (!this.initConfig.cleStockagePageFrom) {
            throwError('Paramètre manquant : cleStockagePageFrom. Est t-il bien présent dans le fichier environnement.ts ?');
        }
        if (!this.environment.pageApresConnexion) {
            throwError('Paramètre manquant : pageApresConnexion. Est t-il bien présent dans le fichier environnement.ts ?');
        }
    }
    postSuccessLogin() {
        // Méthode à surcharger permettant d'ajouter un comportement après l'authentification
        console.log('Dans DefaultSuccessLoginHandler');
    }
}
DefaultSuccessLoginHandler.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: DefaultSuccessLoginHandler, deps: [{ token: PrismeAngularConfiguration }, { token: i1.PrismeAngularInitConfig }], target: i0.ɵɵFactoryTarget.Injectable });
DefaultSuccessLoginHandler.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: DefaultSuccessLoginHandler });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: DefaultSuccessLoginHandler, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i2.PrismeAngularConfiguration, decorators: [{
                    type: Inject,
                    args: [PrismeAngularConfiguration]
                }] }, { type: i1.PrismeAngularInitConfig }]; } });
