// @acoss/prisme-angular-intranet/esm2020/lib/interceptor/prisme-bearer-interceptor.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../service/login.service";
/**
 * Intercepteur qui ajoute à toutes les requêtes le jeton dans l'entête Bearer
 * Ne fait rien si l'utilisateur n'est pas connecté
 */
export class PrismeBearerInterceptor {
    constructor(loginService) {
        this.loginService = loginService;
    }
    intercept(req, next) {
        // Si l'utilisateur est connecté et que l'on a pas déjà un jeton ajouté au moment de la création de la requête
        if (this.loginService.isConnecte() && !this.hasBearerToken(req)) {
            // Ajout de l'access token
            return next.handle(this.ajoutJeton(req));
        }
        // Sinon, rien à faire
        return next.handle(req);
    }
    // Ajoute le jeton dans l'entête Bearer
    ajoutJeton(req) {
        const headers = {
            Authorization: 'Bearer ' + this.loginService.getAccessTokenBack()
        };
        return req.clone({ setHeaders: headers });
    }
    /**
     * Vérifie si on a déjà le jeton d'authentification dans les entêtes de la requête
     * @param req la requête http
     */
    hasBearerToken(req) {
        if (req.headers.has('Authorization')) {
            const headers = req.headers.getAll('Authorization');
            if (headers.find(value => value.startsWith('Bearer ') === true)) {
                return true;
            }
        }
        return false;
    }
}
PrismeBearerInterceptor.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeBearerInterceptor, deps: [{ token: i1.LoginService }], target: i0.ɵɵFactoryTarget.Injectable });
PrismeBearerInterceptor.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeBearerInterceptor });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeBearerInterceptor, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.LoginService }]; } });
