// @acoss/prisme-angular-intranet/esm2020/lib/prisme-angular-init.config.mjs
import { Inject, Injectable, InjectionToken } from '@angular/core';
import * as i0 from "@angular/core";
export const CLE_STOCKAGE = new InjectionToken('CLE_STOCKAGE');
export const UTILISER_LOCAL_STORAGE_TOKEN = new InjectionToken('UTILISER_LOCAL_STORAGE_TOKEN');
export class PrismeAngularInitConfig {
    constructor(cleStockage, utiliserLocalStorageToken) {
        if (cleStockage) {
            this.setCleStockage(cleStockage);
        }
        else {
            this.setCleStockage('defaultPrismeStorageKey');
        }
        if (utiliserLocalStorageToken) {
            this.utiliserLocalStorage = utiliserLocalStorageToken;
        }
        else {
            this.utiliserLocalStorage = false;
        }
    }
    setCleStockage(idCleStockage) {
        this.idCleStockage = idCleStockage;
        this.cleStockageInfosUtilisateur = this.idCleStockage + '_User';
        this.cleStockageDernierNonce = this.idCleStockage + '_Nonce';
        this.cleStockageDernierState = this.idCleStockage + '_LastState';
        this.cleStockageStateUtilise = this.idCleStockage + '_UsedState';
        this.cleStockageEtatRafraichissement = this.idCleStockage + '_RefreshState';
        this.cleStockagePageFrom = this.idCleStockage + '_PageFrom';
    }
}
PrismeAngularInitConfig.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularInitConfig, deps: [{ token: CLE_STOCKAGE }, { token: UTILISER_LOCAL_STORAGE_TOKEN }], target: i0.ɵɵFactoryTarget.Injectable });
PrismeAngularInitConfig.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularInitConfig });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularInitConfig, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [CLE_STOCKAGE]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [UTILISER_LOCAL_STORAGE_TOKEN]
                }] }]; } });
