// @acoss/prisme-angular-intranet/esm2020/lib/prisme-angular-init.module.mjs
import { CommonModule } from '@angular/common';
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { StorageServiceModule } from 'ngx-webstorage-service';
import { PrismeBearerInterceptor } from './interceptor/prisme-bearer-interceptor';
import { CLE_STOCKAGE, PrismeAngularInitConfig, UTILISER_LOCAL_STORAGE_TOKEN } from './prisme-angular-init.config';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * Function permettant d'initiliser la configuration de prismeAngularInitModule
 *
 * @export
 * @param  cleStockage : clé de stockage utilisé dans le sessionStorage
 * @returns la classe PrismeAngularInitConfig instanciée
 */
/**
 * Module permettant d'initialiser la partie minimale de prisme.
 * Il permet entre d'initialiser les modules nécessaires au bearerInterceptor
 * Il prend en paramètre optionnel la clé de stockage
 * @export
 */
export class PrismeAngularInitModule {
    static forRoot(cleStockage, utiliserLocalStorage) {
        return {
            ngModule: PrismeAngularInitModule,
            providers: [PrismeAngularInitConfig, { provide: CLE_STOCKAGE, useValue: cleStockage },
                { provide: UTILISER_LOCAL_STORAGE_TOKEN, useValue: utiliserLocalStorage }]
        };
    }
}
PrismeAngularInitModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularInitModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PrismeAngularInitModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularInitModule, imports: [CommonModule,
        HttpClientModule,
        StorageServiceModule, i1.RouterModule] });
PrismeAngularInitModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularInitModule, providers: [
        { provide: HTTP_INTERCEPTORS, useClass: PrismeBearerInterceptor, multi: true },
    ], imports: [CommonModule,
        HttpClientModule,
        StorageServiceModule,
        RouterModule.forRoot([], { relativeLinkResolution: 'legacy' })] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularInitModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule,
                        HttpClientModule,
                        StorageServiceModule,
                        RouterModule.forRoot([], { relativeLinkResolution: 'legacy' })],
                    providers: [
                        { provide: HTTP_INTERCEPTORS, useClass: PrismeBearerInterceptor, multi: true },
                    ]
                }]
        }] });
