// @acoss/prisme-angular-intranet/esm2020/lib/prisme-angular.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { DefaultFailureLoginHandler } from './handler/default-failure-login.handler';
import { DefaultSuccessLoginHandler } from './handler/default-success-login.handler';
import { PRISME_SUCCESS_LOGIN_HANDLER, PRISME_FAILURE_LOGIN_HANDLER, PrismeAngularConfiguration } from './prisme-angular.config';
import { CallbackComponent } from './component/callback/callback.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * Ce module permet de charger la configuration de prisme.Il va de paire avec le module
 * PrismeAngularInitConfiguration
 *
 * @export
 */
export class PrismeAngularModule {
}
PrismeAngularModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PrismeAngularModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularModule, declarations: [CallbackComponent], imports: [CommonModule, i1.RouterModule], exports: [CallbackComponent] });
PrismeAngularModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularModule, providers: [
        // L'application appelante doit fournir un prismeAngularConfiguration
        { provide: PrismeAngularConfiguration, useExisting: 'PrismeAngularConfiguration' },
        // Si l'application appelante ne surcharge par le success andler, alors on utilise celui par défaut.
        { provide: PRISME_SUCCESS_LOGIN_HANDLER, useClass: DefaultSuccessLoginHandler },
        // Si l'application appelante ne surcharge par le failure handler, alors on utilise celui par défaut.
        { provide: PRISME_FAILURE_LOGIN_HANDLER, useClass: DefaultFailureLoginHandler },
    ], imports: [CommonModule,
        RouterModule.forRoot([], { relativeLinkResolution: 'legacy' })] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: PrismeAngularModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule,
                        RouterModule.forRoot([], { relativeLinkResolution: 'legacy' })],
                    declarations: [CallbackComponent],
                    exports: [CallbackComponent],
                    providers: [
                        // L'application appelante doit fournir un prismeAngularConfiguration
                        { provide: PrismeAngularConfiguration, useExisting: 'PrismeAngularConfiguration' },
                        // Si l'application appelante ne surcharge par le success andler, alors on utilise celui par défaut.
                        { provide: PRISME_SUCCESS_LOGIN_HANDLER, useClass: DefaultSuccessLoginHandler },
                        // Si l'application appelante ne surcharge par le failure handler, alors on utilise celui par défaut.
                        { provide: PRISME_FAILURE_LOGIN_HANDLER, useClass: DefaultFailureLoginHandler },
                    ]
                }]
        }] });
