// @acoss/double-liste-angular-intranet/esm2020/lib/double-liste.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DoubleListeComponent } from './double-liste.component';
import { ReactiveFormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
export class DoubleListeModule {
}
DoubleListeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: DoubleListeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DoubleListeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.1.0", ngImport: i0, type: DoubleListeModule, declarations: [DoubleListeComponent], imports: [CommonModule,
        ReactiveFormsModule], exports: [DoubleListeComponent] });
DoubleListeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: DoubleListeModule, imports: [CommonModule,
        ReactiveFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: DoubleListeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ReactiveFormsModule
                    ],
                    declarations: [DoubleListeComponent],
                    exports: [DoubleListeComponent]
                }]
        }] });
