// @hetznercloud/ngx-translate-mock/esm2020/lib/ngx-translate-mock.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject, of, Subject } from 'rxjs';
import { TRANSLATED_STRING } from './util';
import * as i0 from "@angular/core";
export class TranslateServiceMock {
    constructor() {
        this.onLangChangeSubject = new Subject();
        this.onTranslationChangeSubject = new Subject();
        this.onDefaultLangChangeSubject = new Subject();
        this.isLoadedSubject = new BehaviorSubject(true);
        this.onLangChange = this.onLangChangeSubject.asObservable();
        this.onTranslationChange = this.onTranslationChangeSubject.asObservable();
        this.onDefaultLangChange = this.onDefaultLangChangeSubject.asObservable();
        this.isLoaded = this.isLoadedSubject.asObservable();
        this.languages = ['de'];
    }
    get(content) {
        return of(TRANSLATED_STRING + content);
    }
    use(lang) {
        this.currentLang = lang;
        this.onLangChangeSubject.next({ lang });
    }
    addLangs(langs) {
        this.languages = [...this.languages, ...langs];
    }
    getBrowserLang() {
        return '';
    }
    getLangs() {
        return this.languages;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    getTranslation() {
        return of({});
    }
    instant(key, interpolateParams) {
        return TRANSLATED_STRING + key.toString();
    }
    setDefaultLang(lang) {
        this.onDefaultLangChangeSubject.next(lang);
    }
}
TranslateServiceMock.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TranslateServiceMock, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
TranslateServiceMock.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TranslateServiceMock });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TranslateServiceMock, decorators: [{
            type: Injectable
        }] });
