// @hetznercloud/ngx-translate-mock/esm2020/lib/ngx-translate-mock.module.mjs
/**
 * Copyright (c) 2022 Hetzner Cloud GmbH
 *
 * SPDX-License-Identifier: MIT
 */
import { NgModule } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { TranslateServiceMock } from './ngx-translate-mock.service';
import { TranslateMockDirective } from './ngx-translate-mock.directive';
import { TranslateMockPipe } from './ngx-translate-mock.pipe';
import * as i0 from "@angular/core";
export const TRANSLATE_SERVICE_MOCK = {
    provide: TranslateService,
    useClass: TranslateServiceMock,
};
export class TranslateMockModule {
}
TranslateMockModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TranslateMockModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TranslateMockModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TranslateMockModule, declarations: [TranslateMockPipe, TranslateMockDirective], exports: [TranslateMockPipe, TranslateMockDirective] });
TranslateMockModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TranslateMockModule, providers: [TRANSLATE_SERVICE_MOCK] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TranslateMockModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TranslateMockPipe, TranslateMockDirective],
                    exports: [TranslateMockPipe, TranslateMockDirective],
                    providers: [TRANSLATE_SERVICE_MOCK],
                }]
        }] });
