// @acoss/prisme-angular-intranet/esm2020/lib/model/infos-utilisateur.mjs
/**
 * Permet d'acceder aux donnée fournie par le service user-info
 */
export class InfosUtilisateur {
    constructor(data) {
        this.dataInfosUtilisateur = data;
    }
    getNom() {
        return this.dataInfosUtilisateur.auser_sn;
    }
    getPrenom() {
        return this.dataInfosUtilisateur.auser_givenname;
    }
    /**
     * Retourne uen chaine contenant le prénom et le nom de l'utilisateur
     */
    getNomComplet() {
        return this.dataInfosUtilisateur.auser_cn;
    }
    getCodeAgent() {
        return this.dataInfosUtilisateur.auser_uid;
    }
    getIdSnv2() {
        return this.dataInfosUtilisateur.auser_idsnv2;
    }
    getCodeOrganisme() {
        return this.dataInfosUtilisateur.auser_codeorg;
    }
    getInterRegion() {
        return this.dataInfosUtilisateur.auser_ircode;
    }
    getAdresseMail() {
        return this.dataInfosUtilisateur.auser_mail;
    }
    getVille() {
        return this.dataInfosUtilisateur.auser_persnomville;
    }
    getAdresseEtablissement() {
        return this.dataInfosUtilisateur.auser_etabaddr;
    }
    /**
     * Retourne la liste des droits de l'utilisateur
     */
    getDroits() {
        return this.dataInfosUtilisateur.access;
    }
    /**
     * Retourne les données fournie par le service user-info en JSon
     */
    getDataInfosUtilisateur() {
        return this.dataInfosUtilisateur;
    }
}
